### PR TITLE
Addition of ICNN Initialization Schemes

### PR DESCRIPTION
**Pull request by bunnech**
In this pull request, we add input convex neural network (ICNN) initialization schemes proposed in [Bunne+(2022)](https://arxiv.org/pdf/2206.14262.pdf). Given a source ($\mu$) and target measure ($\nu$), one aims at learning a map $T_\theta$ between $\mu$ and $nu$ parameterized via the gradient of an ICNN. 

To facilitate training, [Bunne+(2022)](https://arxiv.org/pdf/2206.14262.pdf) propose to initialize the ICNN via different initialization schemes: 
(i.) The `identity` function, i.e., before training (and without a pretraining requirement), the gradient of the ICNN maps its input measure onto itself. 
(ii.) The `gaussian` approximation initializes the ICNN in such a way that the gradient of the ICNN maps a measure onto the Gaussian approximation of the opposite measure (utilizing closed-form results for Gaussian optimal transport maps).

This pull request also updates the existing `NeuralDual` to operate on the new ICNN initialization and a notebook comparing both initialization schemes.

### CODE REVIEW

File: ott/core/icnn.py [hunk 1]
```diff
@@ -88,66 +43,142 @@ class ICNN(nn.Module):
       `jax.nn.initializers.normal`).
     act_fn: choice of activation function used in network architecture
       (needs to be convex, default: `nn.leaky_relu`).
+    pos_weights: choice to enforce positivity of weight or use regularizer.
+    gaussian_maps: data inputs of source and target measures for
```
Comment (marcocuturi) [inline ott/core/icnn.py:48]: maybe worth defining types for this

File: ott/core/icnn.py [hunk 2]
```diff
@@ -88,66 +43,142 @@ class ICNN(nn.Module):
...
 
   dim_hidden: Sequence[int]
-  init_std: float = 0.1
+  init_std: float = 1e-1
   init_fn: Callable = jax.nn.initializers.normal
-  act_fn: Callable = nn.leaky_relu
+  act_fn: Callable = nn.relu
   pos_weights: bool = True
+  dim_data: int = 2
+  gaussian_map: Tuple = None
```
Comment (marcocuturi) [inline ott/core/icnn.py:59]: same, maybe define what's in the Tuple ? it's also only 2 elements, right?

File: ott/core/icnn.py [hunk 3]
```diff
@@ -88,66 +43,142 @@ class ICNN(nn.Module):
...
-
-    w_zs = []
+    self.num_hidden = len(self.dim_hidden)
 
     if self.pos_weights:
-      Dense = PositiveDense
+      hid_dense = PositiveDense
+      # this function needs to be the inverse map of function
+      # used in PositiveDense layers
+      rescale = lambda x: jnp.log(jnp.exp(x) - 1)
```
Comment (marcocuturi) [inline ott/core/icnn.py:68]: I think this should come from (and be encapsulated within) the PositiveDense layer. Here is what I would do: 
- add in `PositiveDense` a function called `rectifier_fn`, by default the softplus.
- add also `inv_rectifier_fn`, by default the sofplusinv.
- add a simple assert like `inv_rectifier_fn(rectifier_fn(0.1))==0.1`, "Make sure both rectifier and inverse are defined properly"
- remove `beta` from `PositiveDense` (this can be defined by the user when passing a different rectifier)
- define `rescale = hid_dense.inv_rectifier_fn` in the code above.

File: ott/core/icnn.py [hunk 4]
```diff
@@ -88,66 +43,142 @@ class ICNN(nn.Module):
...
-    z = jnp.multiply(z, z)
-
-    for Wz, Wx in zip(self.w_zs[:-1], self.w_xs[1:-1]):
-      z = self.act_fn(jnp.add(Wz(z), Wx(x)))
-    y = jnp.add(self.w_zs[-1](z), self.w_xs[-1](x))
-
-    return jnp.squeeze(y)
+    for i in range(self.num_hidden + 2):
+      # applies quadratic transform on data
+      # currently this is only (I x + 0), should be T(x) = 2 A x + 2 b
```
Comment (marcocuturi) [inline ott/core/icnn.py:177]: i think this comment can be removed now

File: ott/core/layers.py [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+        inputs, kernel, (((inputs.ndim - 1,), (0,)), ((), ())),
+        precision=self.precision)
+    if self.use_bias:
+      bias = self.param('bias', self.bias_init, (self.dim_hidden,))
+      bias = jnp.asarray(bias, self.dtype)
+      y = y + bias
+    return y
+
+
+class PosDefDense(nn.Module):
```
Comment (marcocuturi) [inline ott/core/layers.py:77]: I think PosDefDense is no longer needed.

**Review by marcocuturi — COMMENTED** [review]
Thanks Charlotte! a few comments to start

File: ott/core/icnn.py [hunk 6]
```diff
@@ -88,66 +43,142 @@ class ICNN(nn.Module):
...
+        1,
+        kernel_init=self.init_fn(self.init_std),
+        bias_init=self.init_fn(self.init_std),
+        use_bias=True,
+      )
     )
     self.w_xs = w_xs
 
+  def compute_gaussian_map(self, inputs):
+    def compute_moments(x, reg=1e-2, sqrt_inv=False):
```
[the remark this answers is not in the document]
Comment (bunnech) [reply ott/core/icnn.py:138]: Sure, done!

File: ott/core/icnn.py (hunk 2 above)
Comment (bunnech) [reply ott/core/icnn.py:59]: Done!

File: ott/core/icnn.py (hunk 4 above)
Comment (bunnech) [reply ott/core/icnn.py:177]: Done!

File: ott/core/layers.py (hunk 5 above)
Comment (bunnech) [reply ott/core/layers.py:77]: Removed.

File: ott/core/icnn.py (hunk 1 above)
Comment (bunnech) [reply ott/core/icnn.py:48]: Done.

File: ott/core/icnn.py (hunk 3 above)
Comment (bunnech) [reply ott/core/icnn.py:68]: Done.

File: ott/core/layers.py [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+    beta: inverse temperature parameter (default: 1.0)
+    dtype: the dtype of the computation (default: float32).
+    precision: numerical precision of computation see `jax.lax.Precision`
+      for details.
+    kernel_init: initializer function for the weight matrix.
+    bias_init: initializer function for the bias.
+  """
+  dim_hidden: int
+  rectifier_fn: Callable = nn.softplus
+  inv_rectifier_fn: Callable = lambda x: jnp.log(jnp.exp(x) - 1)
```
Comment (marcocuturi) [inline ott/core/layers.py:45]: i think the assert that the rectifier and inv_rectifier match is missing

File: ott/core/layers.py [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+      inputs: The nd-array to be transformed.
+    Returns:
+      The transformed input.
+    """
+    inputs = jnp.asarray(inputs, self.dtype)
+    kernel = self.param(
+        'kernel', self.kernel_init, (inputs.shape[-1], self.dim_hidden)
+    )
+    scaled_kernel = self.beta * kernel
+    kernel = jnp.asarray(1 / self.beta * nn.softplus(scaled_kernel), self.dtype)
```
Comment (marcocuturi) [inline ott/core/layers.py:68]: this should call rectifier_fn and remove beta

File: ott/core/layers.py (hunk 7 above)
Comment (bunnech) [reply ott/core/layers.py:45]: Moved it.

File: ott/core/layers.py (hunk 8 above)
Comment (bunnech) [reply ott/core/layers.py]: Done.

File: docs/notebooks/icnn_inits.ipynb [hunk 9]
```diff
@@ -0,0 +1,561 @@
...
+   "metadata": {},
+   "source": [
+    "# Initialization Schemes for Input Convex Neural Network"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "As input convex neural networks (ICNN) are notoriously difficult to train ([Richter-Powell et al., 2021](https://arxiv.org/pdf/2111.12187.pdf), [Bunne et al. (2022)](https://arxiv.org/pdf/2206.14262.pdf) propose to use closed-form solutions between Gaussian approximations to derive relevant parameter initializations for ICNNs.\n",
```
Comment (marcocuturi) [inline docs/notebooks/icnn_inits.ipynb:14]: left but no right parenthesis for Richter-Powell

File: docs/notebooks/icnn_inits.ipynb [hunk 10]
```diff
@@ -0,0 +1,561 @@
...
+    "In particular, [Bunne et al. (2022)](https://arxiv.org/pdf/2206.14262.pdf) address the problem of initializing the parameters of ICNNs to ensure their gradient evaluated at every point is (initially) meaningful in the context of OT, namely that it is able to approximately map a measure $\\nu$ into a target measure $\\nu$. The initializers build heavily on closed-form solutions available for Gaussian approximations of measures ([Gelbrich, 1990](https://onlinelibrary.wiley.com/doi/abs/10.1002/mana.19901470121)).\n",
```
Comment (marcocuturi) [inline docs/notebooks/icnn_inits.ipynb:15]: \nu appears twice.

File: docs/notebooks/icnn_inits.ipynb (hunk 9 above)
Comment (bunnech) [reply docs/notebooks/icnn_inits.ipynb]: Done.

File: docs/notebooks/icnn_inits.ipynb (hunk 10 above)
Comment (bunnech) [reply docs/notebooks/icnn_inits.ipynb]: Done.

File: docs/notebooks/icnn_inits.ipynb [hunk 11]
```diff
@@ -0,0 +1,557 @@
+{
```
Comment (marcocuturi) [inline docs/notebooks/icnn_inits.ipynb:1]: <p>As input convex neural networks (ICNN) are notoriously difficult to train (<a href="https://arxiv.org/pdf/2111.12187.pdf" target="_blank">Richter-Powell et al., 2021</a>), <a href="https://arxiv.org/pdf/2206.14262.pdf" target="_blank">Bunne et al. (2022)</a> propose to use closed-form solutions between Gaussian approximations to derive relevant parameter initializations for ICNNs: they propose, given two measures µ&nbsp;and ν, to initialize ICNN parameters so that they are (initially) meaningful in the context of OT, namely that its gradient is able to approximately map source measure&nbsp;µ&nbsp;onto&nbsp;target ν. These initializations rely on closed-form solutions available for Gaussian measures (<a href="https://onlinelibrary.wiley.com/doi/abs/10.1002/mana.19901470121" target="_blank">Gelbrich, 1990</a>). In this notebook, we introduce the <em>identity</em> and <em>Gaussian approximation</em>-based initialization schemes, and illustrate how they can be used within the <code>OTT</code> library and its <code>ICNN</code>-based <code>NeuralDual</code> module.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ott-jax/ott/pull/90/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (marcocuturi) [inline docs/notebooks/icnn_inits.ipynb:1]: <p>remove</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ott-jax/ott/pull/90/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (marcocuturi) [inline docs/notebooks/icnn_inits.ipynb:1]: <p>mention those helper functions are (after putting <code>plot_ot_map</code> elsewhere) are useful to train solvers?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ott-jax/ott/pull/90/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (marcocuturi) [inline docs/notebooks/icnn_inits.ipynb:1]: <p>if this is only used for plots, maybe introduce right before plots?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ott-jax/ott/pull/90/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (marcocuturi) [inline docs/notebooks/icnn_inits.ipynb:1]: <p>is this used? ctrl+F can't find it</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ott-jax/ott/pull/90/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (marcocuturi) [inline docs/notebooks/icnn_inits.ipynb:1]: <p>the <code>NeuralDual</code> ... ? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ott-jax/ott/pull/90/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

Comment (marcocuturi) [inline docs/notebooks/icnn_inits.ipynb:1]: <ul><li>same, it feels <code>NeuralDual</code> should be a solver or something. </li><li>For more details ..., please look at the documentation (with link)</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ott-jax/ott/pull/90/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='19'/>

Comment (marcocuturi) [inline docs/notebooks/icnn_inits.ipynb:1]: <p>thus no further ... ?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ott-jax/ott/pull/90/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='20'/>

Comment (marcocuturi) [inline docs/notebooks/icnn_inits.ipynb:1]: <p>The Gaussian approximation-based initialization schemes require samples from both, source and target distributions, in order to initialize the ICNNs with linear factors and means, as detailed in <a href="https://arxiv.org/pdf/2206.14262.pdf" target="_blank">Bunne et al. (2022)</a>.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ott-jax/ott/pull/90/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='28'/>

Comment (bunnech) [reply docs/notebooks/icnn_inits.ipynb:1]: <p>?</p>

Comment (bunnech) [reply docs/notebooks/icnn_inits.ipynb:1]: <p>Removed.</p>

Comment (bunnech) [reply docs/notebooks/icnn_inits.ipynb:1]: <p>See comment above.</p>

Comment (bunnech) [reply docs/notebooks/icnn_inits.ipynb:1]: <p>Removed.</p>

Comment (bunnech) [reply docs/notebooks/icnn_inits.ipynb:1]: <p>Introduced it differently.</p>

Comment (bunnech) [reply docs/notebooks/icnn_inits.ipynb:1]: <p>Changed this.</p><p> </p>

Comment (marcocuturi) [reply docs/notebooks/icnn_inits.ipynb:1]: <p>"in the context of the <code>NeuralDual</code> to compute" still sounds like an incomplete sentence.</p>

Comment (bunnech) [reply docs/notebooks/icnn_inits.ipynb:1]: :dizzy_face:

Comment (michalk8) [reply docs/notebooks/icnn_inits.ipynb:1]: <p>I'd put them before the <code>ToyDataset</code> as not to break the flow during the reading. </p><p> </p>

Comment (bunnech) [reply docs/notebooks/icnn_inits.ipynb:1]: This is what I had before, Marco disagreed. I will move them above again, but I would be very happy if this is it now.

File: docs/notebooks/icnn_inits.ipynb [hunk 12]
```diff
@@ -0,0 +1,520 @@
...
```
Comment (michalk8) [inline docs/notebooks/icnn_inits.ipynb:1]: <p>For consistency with most of the tutorials, would not capitalize every noun (only other tutorial that does it is <code>ICNN Dual Solver </code>)</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ott-jax/ott/pull/90/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (michalk8) [inline docs/notebooks/icnn_inits.ipynb:1]: <p>Unnecessary braces in <code>(dataloader_source, ..., )</code> </p><p>In general, would prefer code cells to be formatted (black is ok if @marcocuturi is ok with it; wasn't able to find <code>yapf</code>-based  formatter for notebooks).</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ott-jax/ott/pull/90/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Comment (michalk8) [inline docs/notebooks/icnn_inits.ipynb:1]: <p>Can we please use scientific notation for floats?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ott-jax/ott/pull/90/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

Comment (bunnech) [reply docs/notebooks/icnn_inits.ipynb:1]: <p>I think I disagree, capitalizing headings is also done in other notebooks. It is also the dominant style in most notebook / articles / submission guides of conferences in ML.</p>
